Add support for the ONNX Runtime Eager Mode backend

abock opened the pull request:
This PR implements the necessary hooks/stubs/enums/etc for complete ONNX Runtime (ORT) Eager Mode integration. The actual extension will live out of tree at https://github.com/pytorch/ort.

We have been [working on this at Microsoft](https://github.com/microsoft/onnxruntime-pytorch/tree/eager-ort/torch_onnxruntime) for the last few months, and are finally ready to contribute the PyTorch core changes upstream (nothing major or exciting, just the usual boilerplate for adding new backends).

The ORT backend will allow us to ferry [almost] all torch ops into granular ONNX kernels that ORT will eagerly execute against any devices it supports (therefore, we only need a single ORT backend from a PyTorch perspective).

abock committed:
Add support for the ONNX Runtime Eager Mode backend

This patch implements the necessary hooks/stubs/enums/etc for complete
ONNX Runtime Eager Mode integration. The actual extension lives out
of tree at https://github.com/pytorch/ort.

The old MSNPU backend/keys is renamed to ORT since the old original
MSNPU extension is dead, but Microsoft is continuing on with PyTorch
support for ONNX Runtime. We've agreed to repurpose the values for
MSNPU as ORT to "give back" a bit of space, literally, instead of
adding a new ORT value.

prasanthpul commented:
@jspisak

jspisak commented:
cc @ailzhang @albanD  - can you take a look?

ailzhang commented:
Hey @abock, thanks for the patch! I have a few quick questions to help me understand the torch_ort integration: 
1. Have you tried using PrivateUse1 keys and is it sufficient for your use case before release?
2. What're the use cases for adding AutogradORT key? 
3. I couldn't find some example code in torch_ort repo, would you mind giving a quick summary of torch_ort integration, like what are the target supported models and what it currently supports (it'll be nice if there're some perf numbers as well :D). Also do you have a planned schedule to release torch_ort to public? 
Thanks a lot!

abock commented:
Hi @ailzhang,

In general, the spirit of this PR is much like XLA, HPU, and others.

1. `PrivateUse1` etc will not be sufficient as we (Microsoft) intend to release our ONNX Runtime (ORT) backend as a package that can be installed alongside an upstream build of PyTorch, much like XLA. Most recently it seems MLC and HPU keys were similarly added, [with HPU being an external and private backend](https://github.com/pytorch/pytorch/pull/57094#issuecomment-836179051).

2. Regarding `AutogradORT`, honestly I am not sure how PyTorch [may] use this. We would like to keep our options with ORT flexible - in theory we could leverage PyTorch's autograd engine, or defer that work to ORT's autograd engine. Our backend currently supports training through PyTorch's autograd engine, but whether or not this key comes into play is unknown to me.

3. We are still working on the shape of the extension repo and integrating some other work from other teams at Microsoft so we are all under a single PyTorch ORT umbrella going forward. Currently, you can see our extension here, which requires a PyTorch built with the patch in this PR: https://github.com/microsoft/onnxruntime-pytorch/tree/eager-ort-staging/src/eager. However, this work will be available at https://github.com/pytorch/ort in the coming weeks.

Thank you!

ailzhang commented:
Hi @abock, we're currently only adding dispatch keys to pytorch/pytorch when the integration is ready to release since we're close to the limit of dispatch keys. Would you mind using PrivateUse1 (and AutogradPrivateUse1) for early development for now? Once it's ready for release we can add the keys pretty quickly (in the mean time we're planning to make it easier to add external backends: less files to change etc). 
`using ORT = c10::DispatchKey::PrivateUse1` should do the work :D. Does this sound good?

ezyang commented:
cc @bdhirsh

abock commented:
Per discussion in Slack and email, I've updated the PR to repurpose/rename the old `MSNPU` keys to `ORT` so we can save (literally) a bit in the dispatch key set, etc. `MSNPU` is not used internally at Microsoft anymore, and ongoing work uses the `ORT` key and further functionality in this PR.

albanD commented on c10/core/DispatchKey.h:
```diff
@@ -222,15 +222,16 @@ enum class DispatchKey : uint8_t {
   AutogradCUDA,
   AutogradXLA,
   AutogradLazy,
+  AutogradORT,
```
Do you have an example of how you're going to use that key?

abock replied:
Not at the moment. I've tabled the work on the ORT side that was investigating this path for now, so I can drop it from this PR. We can revisit in another iteration if we need it or something similar.

albanD commented on c10/core/DispatchKey.h:
```diff
@@ -59,8 +59,8 @@ enum class DispatchKey : uint8_t {
-  MSNPU, // unused externally, but tested at
-  // test/cpp_extensions/msnpu_extension.cpp
+  ORT, // ONNX Runtime, lives out of tree at https://github.com/pytorch/ort
+  // test/cpp_extensions/ort_extension.cpp
```
nit can you add here: "and for extension testing in core in:"

albanD commented on test/test_gen_backend_stubs.py:
```diff
@@ -155,7 +155,7 @@ def test_backend_has_no_autograd_key_but_provides_entries(self):
         yaml_str = '''\
 backend: XLA
-cpp_namespace: torch_msnpu
+cpp_namespace: torch_xla
```
Why not vulkan for these ones?

abock replied:
Because `backend: XLA`? I am not sure why this was ever `torch_msnpu`, but changing to `torch_xla` seemed to make the most sense here?

albanD commented on torch/_utils.py:
```diff
@@ -185,6 +185,12 @@ def _rebuild_mlc_tensor(data, dtype, device, requires_grad):
     return tensor
 
 
+def _rebuild_ort_tensor(data, dtype, device, requires_grad):
```
Why not just re-use the function from above?

abock replied:
Copy pasta because no one else decided to reuse previous copies of the same logic either (looking at `_rebuild_mlc_tensor` and `_rebuild_xla_tensor`).

I'd be happy to refactor the three helper functions, but I generally go for "be as least disruptive as possible" and "follow the coding conventions closest to your change" on codebases to which I am not a primary contributor, which is what I did here 😄

abock replied:
I've added a patch to this PR that consolidates these three identical functions just in case we want to do that now. If not I can drop the patch or open a subsequent PR.

albanD commented on torch/_tensor.py:
```diff
@@ -60,7 +60,7 @@ def __deepcopy__(self, memo):
             # doesn't work because of
             # https://github.com/pytorch/pytorch/issues/47442
-            if self.is_sparse or self.device.type in ['xla', 'mlc', 'meta']:
+            if self.is_sparse or self.device.type in ['xla', 'mlc', 'ort', 'meta']:
```
ort doesn't have the concept of storage?

abock replied:
Similar to XLA, we have not exposed the storage behind the `OrtValue` that backs the `ORTTensorImpl` through `c10::TensorImpl::storage()`, etc.

albanD commented on aten/src/ATen/detail/ORTHooksInterface.h:
```diff
@@ -0,0 +1,33 @@
+#include <c10/util/Exception.h>
+#include <c10/util/Registry.h>
+
+constexpr const char* ORT_HELP = " ONNX Runtime (ORT) support loads dynamically via 'import torch_ort'.";
```
Would this be helpful if it was more guided: "You need to 'import torch_ort' to be able to use the 'ort' device in pytorch" ?

ankurvdev committed:
THPDevice_New, THPDevice_init: mark with TORCH_API

These need to be exported to work on Windows.

abock committed:
_tensor.py: refactor and consolidate XLA/ORT/MLC rebuild functions

Consolidate the following functions for rebuilding tensors against
devices that do not support storage:

- _rebuild_xla_tensor
- _rebuild_mlc_tensor
- _rebuild_ort_tensor

Into a single function:

- _rebuild_device_tensor_from_numpy

albanD commented on c10/core/DispatchKey.h:
```diff
@@ -59,8 +59,15 @@ enum class DispatchKey : uint8_t {
-  // test/cpp_extensions/msnpu_extension.cpp
+
+  // ONNX Runtime, lives out of tree at https://github.com/pytorch/ort and
+  // https://github.com/microsoft/onnxruntime, and also used to test general
```
nit: "and is also"

albanD commented on c10/core/DispatchKey.h:
```diff
@@ -59,8 +59,15 @@ enum class DispatchKey : uint8_t {
+  // ONNX Runtime, lives out of tree at https://github.com/pytorch/ort and
+  // https://github.com/microsoft/onnxruntime, and also used to test general
+  // backend/extension machinery in the core. cf:
+  // - test/cpp_extensions/ort_extension.cpp
```
Thanks for updating the list here!

albanD commented on c10/core/DispatchKey.h:
```diff
@@ -222,8 +229,8 @@ enum class DispatchKey : uint8_t {
 
-  // Currently we only have backend-specific autograd keys for CPU/CUDA/XLA and
-  // reserved user-defined backends. All other in-tree backends share the
+  // Currently we only have backend-specific autograd keys for CPU/CUDA/XLA/ORT
```
nit this ORT should be removed no?

albanD commented on c10/core/DispatchKeySet.cpp:
```diff
@@ -19,6 +19,7 @@ constexpr DispatchKeySet backend_dispatch_keyset = autogradother_backends |
         DispatchKey::PrivateUse3,
         DispatchKey::MLC,
         DispatchKey::HPU,
+        DispatchKey::ORT,
```
The fact that MSNPU was not here was a bug?

abock replied:
Yes

albanD commented on torch/_utils.py:
```diff
@@ -173,13 +173,7 @@ def _rebuild_sparse_tensor(layout, data):
-
-
-def _rebuild_mlc_tensor(data, dtype, device, requires_grad):
+def _rebuild_device_tensor_from_numpy(data, dtype, device, requires_grad):
```
code removal 🎉 

For BC reasons, could you add
```python
# Should not be used, only here to be able to load Tensors serialized with older versions of pytorch
_rebuild_xla_tensor = _rebuild_device_tensor_from_numpy
_rebuild_mlc_tensor = _rebuild_device_tensor_from_numpy
```
Python serialization saves the function name and location. So old checkpoint that rely on these functions need it to still exist to be loadable.

albanD reviewed (commented):
The PR looks great!
Thanks for taking the time to do all the code cleanup.

Could you fix CI and then we can land this!

albanD reviewed (approved):
LGTM
Thanks!

abock commented:
Looks like at least one of the failing tests is due to a timeout.

albanD commented:
Yes OSS CI looks good.
This is taking a bit longer than usual because I have to manually update some internal build files and some internal CI is broken.
